### ISSUE TITLE
How to invoke LLVM's whole-program devirtualization feature in bitcode generation using wllvm

### ISSUE THREAD

**Issue by hongkedavid** [travitch/whole-program-llvm#96]
I am trying to build a whole-program bitcode using wllvm for a large C++ project with many source files (configured by bazel build system using CROSSTOOL) and also need to invoke LLVM's whole-program devirtualization feature (_whole-program-vtables_) in the bitcode generation. For a single source file, I can invoke the whole-program devirtualization feature by `clang -c -emit-llvm -flto -fwhole-program-vtables *.cpp -o *.bc` (_lto_ needs to be invoked when invoking _whole-program-vtables_ feature).

However, when I pass the compiler flag _-flto_ and _-fwhole-program-vtables_ into CROSSTOOL and invoke clang and wllvm to generate the whole-program bitcode, a lot of warning outputs `WARNING:Cannot attach bitcode path to "*.o of type UNKNOWN"` are generated and the bitcode generation fails. I checked wllvm's source code and it seems that this warning is related to https://github.com/travitch/whole-program-llvm/blob/474ff73744024a6747d997b2aec99ea6705769e5/wllvm/compilers.py#L115

Is there any clue why this happens? Did I make any wrong configuration? Is there any way to invoke LLVM's whole-program devirtualization feature (_whole-program-vtables_) in wllvm? Thank you for your input.

**Comment by ianamason** [comment]
`wllvm`  invokes `clang` with `-emit-llvm` behind the scenes. So adding flags in that step is doable, but I don't think anyone has done it before. Your comment 
```
CROSSTOOL and invoke clang and wllvm to generate the whole-program bitcode
```
worries me a bit, if you are using `wllvm` then you should not be using `clang` directly.

I am happy to help but we will need to start with more details. Can you point me to tar ball
you are trying to build?

P.S. you really should be using `gllvm`, but we can cross that bridge when we come to it.

**Comment by hongkedavid** [comment]
I appreciate your quick response and help! @ianamason 
Here is a documentation for accessing the target project and my current way of generating the whole-program bitcode for it. 
To correct my previous description, I was invoking `wllvm` to generate the whole-program bitcode. You can check [https://github.com/hongkedavid/apollo_llvm/tree/master/wllvm](the configuration) how I invoke wllvm in the project's bazel build system.

**Comment by ianamason** [comment]
So lets back up a little.

Have you managed to compile it (and produce bitcode) without the  added complication 
of the trying the devirtualization feature?

I am not very familiar with bazel etc, and use vagrant rather than docker usually.

I would much rather start with a small example that is not part of a humongous build system.

There is no problem using `wllvm` or `wllvm++` for large builds, we have built the linux kernel with it.

**Comment by hongkedavid** [comment]
> Have you managed to compile it (and produce bitcode) without the added complication
> of the trying the devirtualization feature?
Yes, when I removed `compiler_flag: "-flto"` in [CROSSTOOL](https://github.com/hongkedavid/apollo_llvm/blob/master/wllvm/CROSSTOOL), the build was successful and I could obtain the whole-program bitcode.

**Comment by ianamason** [comment]
Excellent, That is good news. My guess is that these two compiler flags should only be in the
step that produces the bitcode, not the step that produces the object file. But currently in
`wllvm` (or `gllvm`) there is no way to do that.

I see no problem doing it.

You are going to have to be my tester. OK?

**Comment by hongkedavid** [comment]
Yes, I am happy to be your tester. Please advise me what to do next. Thanks!

**Comment by ianamason** [comment]
OK so while I am paging in the code, you should move your build system to using my fork:

```
git clone https://github.com/SRI-CSL/whole-program-llvm.git
cd whole-program-llvm
make develop
```
should replace the step
```
pip install wllvm
```
Make sure that works without the new virtualization flags.

**Comment by hongkedavid** [comment]
Got it. I am trying your fork and let you know how it works.

**Comment by ianamason** [comment]
How are you doing?

OK  now you should be able to  do a `git pull` in your `whole-program-llvm` repository
and then add these to your wllvm and wllvm++ environments.

```
LLVM_BITCODE_GENERATION_FLAGS="-flto -fwhole-program-vtables"
```

**Comment by hongkedavid** [comment]
Just finished building the bitcode without the virtualization flags. It works. 
I am pulling your update and run wllvm again to see if it also works. Thanks.

**Comment by hongkedavid** [comment]
With the virtualization flags added to the wllvm and wllvm++ environments, wllvm can also generate the bitcode now! I have tested it in several modules of the project and they all worked.
I really appreciate your quick response and support of this new feature. It greatly helps my research! 
I am closing the issue now.

**Comment by ianamason** [comment]
Your build times would be halved if you used gllvm. Let me know if you are interested.
It's a trivial switch.

**Comment by hongkedavid** [comment]
Yes, I am interested in trying gllvm. Is there any documentation to gain some high-level understanding of gllvm and also how to install and configure it.

**Comment by ianamason** [comment]
https://github.com/SRI-CSL/gllvm has a good README.  You will want to wait until I port
yesterday's fix, which I will do this afternoon. I will let you know.

**Comment by ianamason** [comment]
OK I have ported the new feature to `gllvm`.
So it should be easy for you to make the switch (I think) you just need to install  `golang`
then do 
```
go get github.com/SRI-CSL/gllvm/cmd/...
```
Those dots are part of the command. All the environment variables are the same.

* `wllvm` should be replaced by `gclang`,
* `wllvm++` should be replaced by `gclang++`, and
* `extract-bc` should be replaced by `get-bc`.

Just like yesterday, you are my tester.

**Comment by hongkedavid** [comment]
Thanks for your instructions. I will test gllvm later this week and get back to you.

**Comment by ianamason** [comment]
Great. Thanks.

**Comment by ianamason** [comment]
Hey @hongkedavid did you ever get around to testing this new feature in gllvm for me?
